Fix default file inclusion of .RESW and image files in .NET projects

rohanp-msft opened the pull request:
The existing solution (added in #771) has an issue: app devs cannot opt out an image or a RESW file from PRI inclusion.

Closes #980 

**How verified (commit 11)**

Setup:
1. Uninstall the existing VSIX.
2. Install WindowsAppSDK.Extension.Experimental.vsix from the output of the build pipeline run.
3. Create a new app using the "WinUI.Desktop.Cs.PackagedApp" template.
4. Install Microsoft.WindowsAppRuntime.1.0.msix manually.

Verification:
(BA below means Build Action)
1. Copy paste a RESW file, a .PNG file, and a .JPG file into the folder containing the .CSPROJ file. Build and run. Examine PRI. Examine bin placing.
2. Opt-out the RESW file and the PNG file. Build and run. Examine PRI. Examine bin placing.
3. Opt-in the files that were opted out. Build and run. Examine PRI. Examine bin placing.
4. Restart VS. Rebuild and run. Examine PRI.
5. Opt-out files. Restart VS. Build. Examine PRI.
6. Set `EnableDefaultPRIResourceItems` to false. BA of RESW file changed to `None`. Remove set. BA changes back to `PRIResource`.
7. Set `EnableDefaultContentItems` to false. BA of image file changed to `None`. Remove set. BA changes back to `Content`. (BA of image file that was `None` already remains None and is `None` after set is removed too.)
8. Set `EnableDefaultItems` to false. All image and RESW files disappeared in Solution Explorer (regardless of their BA). Remove set. All files come back.

**Note**:
- *Issue 1*: There is an issue I found for which I'll create a tracking item and link to this PR:
  - Let's say you create a new app. You build and run it. That generates a resources.pri.
  - You then copy-paste some resource files into the project folder. The Build Action of these files will be set correctly by default (which is the point of the change).
  - You then hit F5.
  - The resources.pri in the build output after the F5 Run doesn't have the new resources. I suspect the PRI is not regenerated.
  - If the app is rebuilt, then the PRI's contents are correct.
- *Issue 2*: This solution does not work for resource files that are added from outside the project folder. For them, manual action is needed to set the Build Action to `PRIResource` or `Content`. I'll create a tracking item for this too and link to this PR.

rohanp-msft commented:
/azp run

rohanp-msft commented:
/azp run

rainersigwald commented on build/NuSpecs/Microsoft.WindowsAppSDK.Foundation.props:
```diff
@@ -8,4 +8,14 @@
+    <EnableCoreMrtTooling Condition="'$(EnableCoreMrtTooling)'=='' and '$(WindowsAppContainer)'!='true' and '$(MSBuildProjectExtension)' != '.wapproj'">true</EnableCoreMrtTooling>
+  </PropertyGroup>
+
+  <Import Project="$(MSBuildThisFileDirectory)Microsoft.Windows.ApplicationModel.Resources.props" Condition="Exists('$(MSBuildThisFileDirectory)Microsoft.Windows.ApplicationModel.Resources.props') and '$(EnableCoreMrtTooling)'=='true'" />
```
Nit: in the case of missing files that should always be in your NuGet package/installed/present, it's usually preferable to avoid the `Exists` condition and just let MSBuild error on the import. Otherwise you have to spend time debugging "the project loaded fine but didn't behave right" and eventually figure out that it was because a file was missing.

```suggestion
  <Import Project="$(MSBuildThisFileDirectory)Microsoft.Windows.ApplicationModel.Resources.props" Condition="'$(EnableCoreMrtTooling)'=='true'" />
```

rohanp-msft replied:
Done!

huichen123 reviewed (approved):
:shipit:

rohanp-msft commented:
@rainersigwald and @huichen123 
When you get some time, would you revisit the change please? I fixed a bug: `EnableDefaultContentItems` should only affect Content. I also added `EnableDefaultPRIResourceItems`, something that doesn't exist currently, so that app devs can switch off default inclusion of PRIResource items, just like they can for Content, Compile, etc.

Disclaimer: Verification of commit 6 is pending. There are no new changes that I think will cause the fix to stop working. I will still do the verification, of course, but I want to be sure that things like the introduction of `EnableDefaultPRIResourceItems` is not unacceptable (if it was, that'll make my time-expensive verification a waste).

rohanp-msft commented:
/azp run

rainersigwald commented on dev/MRTCore/packaging/MrtCore.props:
```diff
@@ -0,0 +1,32 @@
+    <None Remove="$(_ReunionImageGlobs)" />
+    <Content Include="$(_ReunionImageGlobs)" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" />
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)'=='.NETCoreApp' and '$(EnableDefaultItems)'=='true' and '$(EnableDefaultPRIResourceItems)'!='false'">
+    <None Remove="**/*.resw" />
+    <PRIResource Include="**/*.resw" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" />
```
I think this makes sense but it did give me pause for a second. Specifically, the idea that `$(EnableDefaultPRIResourceItems)` doesn't control whether the files get globbed, just what item they get globbed into, is a bit surprising. Is that preferable to always removing resw from None (when using your package) and only conditionally including it?

I am fine with either way but want to make sure you think about it explicitly.

dsplaisted replied:
I think leaving resw files out of `None` would more closely match what we now do in the .NET SDK.  If you turn `EnableDefaultCompileItems` off, the source files will not be part of `None`.  Originally the behavior was different, but we changed it to address [some bugs](https://github.com/dotnet/sdk/issues/1157) in Visual Studio behavior.

So I would lean towards removing `*.resw` from `None` regardless of whether `EnableDefaultPRIResourceItems` is set.  It's up to you, but if you want to go the other way I'd suggest testing what happens if you set the property to false and then add a resw file to the project in Visual Studio, both via the new item template as well as add / existing item.

rohanp-msft replied:
Went with Daniel's suggestion to remove `*.resw` from `None` regardless of whether `EnableDefaultPRIResourceItems` is set.

rainersigwald reviewed (approved):
> I want to be sure that things like the introduction of `EnableDefaultPRIResourceItems` is not unacceptable

I think this is a very reasonable extension of the current pattern. With MSBuild there's always a risk of name collision but this is very specific and doesn't appear to be currently used.

@dsplaisted do you have any concerns from the SDK perspective?

dsplaisted commented on build/NuSpecs/Microsoft.WindowsAppSDK.Foundation.props:
```diff
@@ -8,4 +8,11 @@
 
+  <!-- At least for now, we want MRT Core to not be used in WAP or UWP projects. There, we want the in-box PRI generation tooling to be used. -->
+  <PropertyGroup>
+    <EnableCoreMrtTooling Condition="'$(EnableCoreMrtTooling)'=='' and '$(WindowsAppContainer)'!='true' and '$(MSBuildProjectExtension)'!='.wapproj'">true</EnableCoreMrtTooling>
```
It seems like this condition won't entirely work now that it's moved to a .props file.  The `EnableCoreMrtTooling` and `WindowsAppContainer` properties wouldn't be set yet.

rohanp-msft replied:
Excellent find...

When I moved the property definition, that didn't occur to me. The import then worked in my verification, so I didn't question it. I see now that the reason it worked is that my conditions check if the properties are not equal to fixed strings. Empty strings will of course satisfy that.

Will fix!

I wish MSBuild logs captured where a property was assigned its value, not just what its value is. That would have helped in figuring out why this import was working.

dsplaisted replied:
[MSBuild binary logs](https://msbuildlog.com) actually do capture the locations where properties are reassigned.  I'm not sure if that tells you where it was first set, but if it is changed then it will show you where.

![image](https://user-images.githubusercontent.com/145043/132929974-31b43313-ea97-4ee5-8714-65b3d368049e.png)

rohanp-msft replied:
That's good to know! I think your suspicion about the first assignment is right. In my case, the property was not reassigned a value and I didn't see the location of the assignment in the logs, just its value.

rainersigwald replied:
> I wish MSBuild logs captured where a property was assigned its value, not just what its value is.

https://github.com/dotnet/msbuild/issues/2711. We need to figure out an efficient way to do it.

dsplaisted reviewed (approved):
The globbing looks good, but it seems like there may be issues with moving the `EnableCoreMrtTooling` logic to .props files.

A possible way to work around this would be to move the logic that sets `EnableCoreMrtTooling` back to the .targets, import `MrtCore.props` unconditionally, and then add a check for `EnableCoreMrtTooling` to all the ItemGroups in `MrtCore.props`.

rohanp-msft commented:
/azp run

huichen123 commented on dev/MRTCore/packaging/MrtCore.props:
```diff
@@ -0,0 +1,37 @@
+    this file. This is correct though because ItemGroup definitions are evaluated in a later pass. See
+    https://docs.microsoft.com/en-us/visualstudio/msbuild/build-process-overview?view=vs-2019#evaluation-phase.
+  -->
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)'=='.NETCoreApp' and '$(EnableCoreMrtTooling)'=='true'">
```
>'$(EnableCoreMrtTooling)'=='true'"> [](http://example.com/codeflow?start=74&length=35)

this is not true

rohanp-msft replied:
Item groups are evaluated in a later pass than properties. Please see https://docs.microsoft.com/en-us/visualstudio/msbuild/build-process-overview?view=vs-2019#evaluation-phase

rohanp-msft commented:
/azp run
